crates/worker/src/partition_processor_manager/snapshot_task.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::path::PathBuf;
+use std::time::SystemTime;
+
+use tracing::{debug, instrument, warn};
+
+use restate_core::worker_api::SnapshotError;
+use restate_partition_store::snapshots::{
+    LocalPartitionSnapshot, PartitionSnapshotMetadata, SnapshotFormatVersion,
+};
+use restate_partition_store::PartitionStoreManager;
+use restate_types::identifiers::{PartitionId, SnapshotId};
+
+/// Creates a partition store snapshot along with Restate snapshot metadata.
+pub struct SnapshotPartitionTask {
+    pub snapshot_id: SnapshotId,
+    pub partition_id: PartitionId,
+    pub snapshot_base_path: PathBuf,
+    pub partition_store_manager: PartitionStoreManager,
+    pub cluster_name: String,
+    pub node_name: String,
+}
+
+impl SnapshotPartitionTask {
+    #[instrument(level = "info", skip_all, fields(snapshot_id = %self.snapshot_id, partition_id = %self.partition_id))]
+    pub async fn run(self) -> Result<PartitionSnapshotMetadata, SnapshotError> {
+        debug!("Creating partition snapshot");
+
+        let result = create_snapshot_inner(
+            self.snapshot_id,
+            self.partition_id,
+            self.partition_store_manager,
+            self.snapshot_base_path,
+            self.cluster_name,
+            self.node_name,
+        )
+        .await;
+
+        result
+            .inspect(|metadata| {
+                debug!(
+                    archived_lsn = %metadata.min_applied_lsn,
+                    "Partition snapshot created"
+                );
+            })
+            .inspect_err(|err| {
+                warn!("Failed to create partition snapshot: {}", err);
+            })
+    }
+}
+
+async fn create_snapshot_inner(
+    snapshot_id: SnapshotId,
+    partition_id: PartitionId,
+    partition_store_manager: PartitionStoreManager,
+    snapshot_base_path: PathBuf,
+    cluster_name: String,
+    node_name: String,
+) -> Result<PartitionSnapshotMetadata, SnapshotError> {
```

Comment:
Was there a reason to make this function not a method of `SnapshotPartitionTask`? If it were, then you wouldn't have to pass in all the parameters explicitly. Instead it could accept `self`.

Reply:
I think this is a leftover from the earlier channels-based communication, will revisit.

Reply:
Moved this and write metadata to methods, also got rid of some clones in the process! 👍